File: Questão11.py
from abc import ABC, abstractmethod

class Funcionario(ABC):
    """Classe abstrata que representa um funcionário."""

    @abstractmethod
    def calcular_salario(self) -> float:
        pass

class FuncionarioHorista(Funcionario):
    """Classe que representa um funcionário horista."""
    
    def __init__(self, nome: str, horas_trabalhadas: float, valor_hora: float) -> None:
        self.nome = nome
        self.horas_trabalhadas = horas_trabalhadas
        self.valor_hora = valor_hora

    def calcular_salario(self) -> float:
        return self.horas_trabalhadas * self.valor_hora

class FuncionarioAssalariado(Funcionario):
    """Classe que representa um funcionário assalariado."""
    
    def __init__(self, nome: str, salario: float) -> None:
        self.nome = nome
        self.salario = salario

    def calcular_salario(self) -> float:
        return self.salario


funcionarios = [
    FuncionarioHorista("Alice", 160, 50),
    FuncionarioAssalariado("Bob", 4000)
]

for funcionario in funcionarios:
    print(f"Salário de {funcionario.nome}: R${funcionario.calcular_salario():.2f}")

File: Questão11.java
abstract class Funcionario {
    public abstract double calcularSalario();
}

class FuncionarioHorista extends Funcionario {
    private String nome;
    private double horasTrabalhadas;
    private double valorHora;

    public FuncionarioHorista(String nome, double horasTrabalhadas, double valorHora) {
        this.nome = nome;
        this.horasTrabalhadas = horasTrabalhadas;
        this.valorHora = valorHora;
    }

    @Override
    public double calcularSalario() {
        return horasTrabalhadas * valorHora;
    }

    public String getNome() {
        return nome;
    }
}

class FuncionarioAssalariado extends Funcionario {
    private String nome;
    private double salario;

    public FuncionarioAssalariado(String nome, double salario) {
        this.nome = nome;
        this.salario = salario;
    }

    @Override
    public double calcularSalario() {
        return salario;
    }

    public String getNome() {
        return nome;
    }
}

public class Main {
    public static void main(String[] args) {
        Funcionario[] funcionarios = {
            new FuncionarioHorista("Alice", 160, 50),
            new FuncionarioAssalariado("Bob", 4000)
        };

        for (Funcionario f : funcionarios) {
            System.out.printf("Salário de %s: R$%.2f%n", f.getNome(), f.calcularSalario());
        }
    }
}
